### DISCUSSION TITLE
Experiment with Tauri

### DISCUSSION THREAD

**Discussion by martpie** [martpie/museeks#639] [Ideas]
> [!IMPORTANT] 
> March 5th 2024: work is in good progress although there are a few important blockers. See progress and what's left here https://github.com/martpie/museeks/pull/753

I've been playing a little bit with [Tauri](https://tauri.studio/) and it seems an early, but very nice solution. It has less APIs but look much more frameworky than Electron (binaries compilation included, etc). 

I don't think we can re-use 1:1 compared with Electron, and we may lose some features along the way, but consuming 5 times less RAM and dividing binary size by 15 seem like awesome things to have.

_edit_:  Please see #753 for the list of leftovers. There are a bunch of broken features, hardly fixable.

~~I am also thinking on switching to something lighter than React. The main reason is I'm going to have to rewrite a good part of the React codebase, so if there's a way to simplify this part as well, why not?~~

---

Early results below:

### 2024-09-05

Most features are working % hard problems to solve. Missing features may be removed.

### 2024-02-22

I have a loading version of Museeks working + library import from UI working. Everything else is broken.

### 2024-02-19

I started experimenting again with Tauri 2.0 beta. My goal is to have a minimal version of Museeks running on Tauri quickly (ID3 scan + DB + basic player), but with many features/integrations broken. I really hope to be able to keep the momentum, this is a fun change.

### 2023-05-26 (yes, one year later)

We'll keep React, changing both is too much work. My strategy is to use the same UI but change the backend, then revamp the UI (#640), then _maybe_, move to Solid (probably not gonna happen)

### 2022-05-28

- Library used to extract metadata from songs is working but has some limitations, currently contributing to it to improve it (PRs to `rust-id3` and forked `audiotags`)
- Tauri is not based on Chromium, but most of the formats we're interested in seem covered
- Solid is great, but some UI components WILL need virtualization for performance reasons

### 2022-05-24 

- Scanning and importing 4000 tracks
  - Before: 3-4mn
  - After: <30 sec
- Loading 4000 tracks
  - Before: couple of seconds
  - After: 180ms

**Comment by binyamin** [comment]
@martpie There's also [Preact](https://preactjs.com/), which tries to expose the same React API in around 3kb.

**Reply by martpie** [reply]
So the main advantage I see with Solid.js is the absence of Virtual DOM. That means the tracks list would be heavily optimized (we currently have to rely on a lot of small tricks to keep the thing performant enough).

I tested with 10'000 tracks on Tauri, and it didn't need to use any kind of optimizations to perform well, so that's a big plus. The downside is the ecosystem is VERY limited.

The UI stack is not set in stone anyway, I am currently mostly experimenting with the Rust back-end, and see if I have all the tools to re-create the library scan, the ID3 tags extraction, the DB management, etc.

**Comment by marcocastignoli** [comment]
There is this now! https://github.com/agmmnn/tauri-controls

**Reply by martpie** [reply]
This might be an excellent workaround waiting for an official solution, thanks for sharing!
